File: app/src/main/java/com/jskhaleel/learntamil/utils/ImageTextReader.java
package com.jskhaleel.learntamil.utils;

import android.graphics.Bitmap;

import com.googlecode.tesseract.android.TessBaseAPI;

/**
 * This class convert the image to text and return the text on image
 */
public class ImageTextReader {

    /**
     * TessBaseAPI instance
     */
    private static volatile TessBaseAPI api;
    //  private static volatile TesseractImageTextReader INSTANCE;

    /**
     * initialize and train the tesseract engine
     *
     * @param path     a path to training data
     * @param language language code i.e. selected by user
     * @return the instance of this class for later use
     */
    public static ImageTextReader geInstance(String path, String language) {
        api = new TessBaseAPI();
        api.init(path, language, TessBaseAPI.OEM_TESSERACT_ONLY);
//        api.setPageSegMode(TessBaseAPI.PageSegMode.PSM_AUTO_OSD);
        return new ImageTextReader();
    }

    /**
     * get the text from bitmap
     *
     * @param bitmap a image
     * @return text on image
     */
    public String getTextFromBitmap(Bitmap bitmap) {
        api.setImage(bitmap);
        String textOnImage;
        try {
            textOnImage = api.getUTF8Text();
        } catch (Exception e) {
            return "Scan Failed: WTF: Must be reported to developer!";
        }
        if (textOnImage.isEmpty()) {
            return "Scan Failed: Couldn't read the image\nProblem may be related to Tesseract or no Text on Image!";
        } else return textOnImage;
    }

    public void clearReader() {
        api.clear();
    }
}

File: app/src/main/java/com/jskhaleel/learntamil/utils/ImageTextReader.kt
package com.jskhaleel.learntamil.utils

import android.graphics.Bitmap
import com.googlecode.tesseract.android.TessBaseAPI

/**
 * This class convert the image to text and return the text on image
 */
class ImageTextReader {
    /**
     * get the text from bitmap
     *
     * @param bitmap a image
     * @return text on image
     */
    fun getTextFromBitmap(bitmap: Bitmap?): String {
        api!!.setImage(bitmap)
        val textOnImage: String = try {
            api!!.utF8Text
        } catch (e: Exception) {
            return "Scan Failed: WTF: Must be reported to developer!"
        }
        return if (textOnImage.isEmpty()) {
            "Scan Failed: Couldn't read the image\nProblem may be related to Tesseract or no Text on Image!"
        } else textOnImage
    }

    fun clearReader() {
        api!!.clear()
    }

    companion object {
        /**
         * TessBaseAPI instance
         */
        @Volatile
        private var api: TessBaseAPI? = null
        //  private static volatile TesseractImageTextReader INSTANCE;
        /**
         * initialize and train the tesseract engine
         *
         * @param path     a path to training data
         * @param language language code i.e. selected by user
         * @return the instance of this class for later use
         */
        fun geInstance(path: String?, language: String?): ImageTextReader {
            api = TessBaseAPI()
            api!!.init(path, language, TessBaseAPI.OEM_TESSERACT_ONLY)
            //        api.setPageSegMode(TessBaseAPI.PageSegMode.PSM_AUTO_OSD);
            return ImageTextReader()
        }
    }
}